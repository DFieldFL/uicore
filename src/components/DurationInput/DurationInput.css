.main {
  position: relative;
  display: inline-flex;
  padding-right: calc(14px + calc(2 * var(--spacing-small)));

  .helpIcon {
    position: absolute;
    right: 0;
    bottom: 0;
    margin: var(--spacing-small);
  }
}
