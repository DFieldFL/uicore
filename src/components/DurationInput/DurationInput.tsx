import React from 'react'

import { TextInput, TextInputProps } from '../TextInput/TextInput'
import { Icon } from '../../icons/Icon'
import { Popover } from '../Popover/Popover'
import { Text } from '../Text/Text'

import css from './DurationInput.css'

export type Units = 'w' | 'd' | 'h' | 'm' | 's' | 'ms'
const TEXT_EXTRACT_REGEX = /(\d+)\s*([a-z]{1,2})/gi
const TEXT_LIMIT_REGEX = /[^0-9wdhms\s]/g
const UNIT_LESS_REGEX = /\d+(?!(ms|s|m|h|d|w|\d))/i
const VALID_SYNTAX_REGEX = /^(\d+w\s*)?(\d+d\s*)?(\d+h\s*)?(\d+m\s*)?(\d+s\s*)?(\d+ms\s*)?$/i

export const UNIT_MULTIPLIIERS: Record<Units, number> = {
  ms: 1,
  s: 1e3,
  m: 1e3 * 60,
  h: 1e3 * 60 * 60,
  d: 1e3 * 60 * 60 * 24,
  w: 1e3 * 60 * 60 * 24 * 7
}

const UNITS_ORDER: Units[] = ['w', 'd', 'h', 'm', 's', 'ms']

export const ALL_UNITS: Units[] = ['w', 'd', 'h', 'm', 's', 'ms']

const UNITS_MAP = new Map([
  ['w', 'weeks'],
  ['d', 'days'],
  ['h', 'hours'],
  ['m', 'minutes'],
  ['s', 'seconds'],
  ['ms', 'milliseconds']
])
/**
 * Converts time string like '1d 2h' to time (in ms)
 */
export function parseStringToTime(str: string): number {
  let time = 0
  /*
      A varibale for tracking if a unit is processed already.
      This will help in avoiding duplicate units.

      For example if a user enters '1w 2d 2w',
      only '1w' will be considred and '2w' will be ignored
    */
  const done: Record<Units, boolean> = {
    w: false,
    d: false,
    h: false,
    m: false,
    s: false,
    ms: false
  }

  /**
   * String.replace calls the callback with all the matching groups
   * as arguments from position 2
   *
   * @param {String} _     The entire string
   * @param {String} n     First capture group, the number in this case
   * @param {String} units Second capture group, the unit in this case
   */
  str.replace(TEXT_EXTRACT_REGEX, (_: string, n: string, units: string) => {
    const unit = units.toLowerCase() as Units

    // check in unit is valid and not already captured
    if (unit in UNIT_MULTIPLIIERS && !done[unit]) {
      // add to the time
      time += parseInt(n, 10) * UNIT_MULTIPLIIERS[unit]
      // set the unit as already captured
      done[unit] = true
    }

    return ''
  })

  return time
}

/**
 * Converts given time (in ms) to text like '2h 30m'
 */
export function timeToDisplayText(time: number): string {
  if (time < 1) return ''

  const str: string[] = []
  let t = time

  UNITS_ORDER.forEach(key => {
    const n = Math.floor(t / UNIT_MULTIPLIIERS[key])
    t = t % UNIT_MULTIPLIIERS[key]

    if (n > 0) {
      str.push(`${n}${key}`)
    }
  })
  return str.join(' ')
}

export const getHelpPopoverContent = (allowedUnits: string[]) => (
  <Text padding="xlarge" style={{ minWidth: '192px' }}>
    You can use:
    <br />
    {allowedUnits.map((unit: string) => {
      return (
        <>
          <br />
          <code>
            <b>{unit}</b>
          </code>
          <span>&nbsp;&nbsp;for {UNITS_MAP.get(unit)}</span>
        </>
      )
    })}
  </Text>
)

export interface DurationInputProps
  extends Omit<TextInputProps, 'value' | 'onChange' | 'valueInTimeFormat' | 'allowedUnits' | 'allowVariables'> {
  value?: number
  // will be string if valueInTimeFormat is passed
  valueInTimeFormat?: string
  onChange?(time: number | string, hasWarning?: boolean): void
  allowedUnits?: string[]
  allowVariables?: boolean
}
export function DurationInput(props: DurationInputProps) {
  const { value, valueInTimeFormat, allowVariables, allowedUnits, onChange, ...rest } = props

  const [text, setText] = React.useState(
    // do not trim any valueInTimeFormat
    !valueInTimeFormat && valueInTimeFormat !== '' ? timeToDisplayText(value || 0).trim() : valueInTimeFormat
  )
  const [showWarning, setShowWarning] = React.useState(false)

  React.useEffect(() => {
    setText(!valueInTimeFormat && valueInTimeFormat !== '' ? timeToDisplayText(value || 0).trim() : valueInTimeFormat)
  }, [value])

  function handleTextChange(e: React.ChangeEvent<HTMLInputElement>) {
    const fieldValue =
      (allowVariables && e.target.value.startsWith('$') && e.target.value) ||
      e.target.value.replace(TEXT_LIMIT_REGEX, '')
    let hasWarning = UNIT_LESS_REGEX.test(fieldValue) || !VALID_SYNTAX_REGEX.test(fieldValue)

    if (allowVariables && e.target.value.startsWith('${') && e.target.value.endsWith('}')) {
      hasWarning = false
    }

    if (allowedUnits) {
      // if limited allowed units, show warning when non-expression value breaks the rule
      const diff = ALL_UNITS.filter(unit => !allowedUnits.includes(unit))
      if (!e.target.value.startsWith('$') && diff.some(unit => e.target.value.includes(unit))) {
        hasWarning = true
      }
    }
    setText(fieldValue)
    setShowWarning(hasWarning)

    // call onChange only when numbers are followed by allowed units
    if (typeof onChange === 'function' && !hasWarning) {
      // should not trim fieldValue for valueInTime
      const time = !valueInTimeFormat && valueInTimeFormat !== '' ? parseStringToTime(fieldValue) : fieldValue
      onChange(time)
    }
  }

  return (
    <div className={css.main}>
      <TextInput
        placeholder="Enter w/d/h/m/s/ms"
        {...rest}
        value={text}
        onChange={handleTextChange}
        rightElement={showWarning ? 'warning-sign' : undefined}
        rightElementProps={{ className: css.warnIcon }}
      />
      <Popover
        wrapperTagName="div"
        className={css.helpIcon}
        content={getHelpPopoverContent(allowedUnits || ALL_UNITS)}
        lazy={true}
        interactionKind="hover"
        position="top"
        usePortal={false}>
        <Icon name="question" size={14} />
      </Popover>
    </div>
  )
}
