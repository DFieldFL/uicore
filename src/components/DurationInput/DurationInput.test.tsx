import React from 'react'
import { render, fireEvent, waitForDomChange } from '@testing-library/react'

import { DurationInput, timeToDisplayText, parseStringToTime, UNIT_MULTIPLIIERS } from './DurationInput'

describe('<DurationInput/> tests', () => {
  describe('component tests', () => {
    test('renders with given value', () => {
      const { container } = render(<DurationInput value={123456789} />)

      expect(container).toMatchSnapshot()
    })

    test('onChange triggers with value (in ms)', () => {
      const onChange = jest.fn()
      const { getByTestId } = render(<DurationInput onChange={onChange} data-testid="input" />)

      fireEvent.change(getByTestId('input'), {
        target: {
          value: '1d 10h 17m 36s 789ms'
        }
      })

      expect(onChange).toHaveBeenCalledWith(123456789)
    })

    test('shows help popover', async () => {
      const { container } = render(<DurationInput value={123456789} />)

      const icon = container.querySelector('.bp3-popover-target')

      fireEvent.mouseOver(icon!)

      await waitForDomChange({ container: document.body })

      expect(container).toMatchSnapshot()
    })
  })

  describe('timeToDisplayText tests', () => {
    test('works with positive', () => {
      expect(timeToDisplayText(UNIT_MULTIPLIIERS.ms)).toBe('1ms')
      expect(timeToDisplayText(UNIT_MULTIPLIIERS.s)).toBe('1s')
      expect(timeToDisplayText(UNIT_MULTIPLIIERS.m)).toBe('1m')
      expect(timeToDisplayText(UNIT_MULTIPLIIERS.h)).toBe('1h')
      expect(timeToDisplayText(UNIT_MULTIPLIIERS.d)).toBe('1d')
      expect(timeToDisplayText(UNIT_MULTIPLIIERS.w)).toBe('1w')
      expect(
        timeToDisplayText(
          2 * UNIT_MULTIPLIIERS.w +
            4 * UNIT_MULTIPLIIERS.d +
            5 * UNIT_MULTIPLIIERS.h +
            15 * UNIT_MULTIPLIIERS.m +
            50 * UNIT_MULTIPLIIERS.s
        )
      ).toBe('2w 4d 5h 15m 50s')
    })

    test('does not works with non-positive values', () => {
      expect(timeToDisplayText(0)).toBe('')
      expect(timeToDisplayText(-1900)).toBe('')
    })
  })

  describe('parseStringToTime tests', () => {
    test('works with simple values', () => {
      expect(parseStringToTime('1ms')).toBe(UNIT_MULTIPLIIERS.ms)
      expect(parseStringToTime('1s')).toBe(UNIT_MULTIPLIIERS.s)
      expect(parseStringToTime('1m')).toBe(UNIT_MULTIPLIIERS.m)
      expect(parseStringToTime('1h')).toBe(UNIT_MULTIPLIIERS.h)
      expect(parseStringToTime('1d')).toBe(UNIT_MULTIPLIIERS.d)
      expect(parseStringToTime('1w')).toBe(UNIT_MULTIPLIIERS.w)
    })

    test('works with complex values', () => {
      expect(parseStringToTime('2w 4d 5h 15m 50s')).toBe(
        2 * UNIT_MULTIPLIIERS.w +
          4 * UNIT_MULTIPLIIERS.d +
          5 * UNIT_MULTIPLIIERS.h +
          15 * UNIT_MULTIPLIIERS.m +
          50 * UNIT_MULTIPLIIERS.s
      )
      expect(parseStringToTime('12w 5d 10h 30m 33s')).toBe(
        12 * UNIT_MULTIPLIIERS.w +
          5 * UNIT_MULTIPLIIERS.d +
          10 * UNIT_MULTIPLIIERS.h +
          30 * UNIT_MULTIPLIIERS.m +
          33 * UNIT_MULTIPLIIERS.s
      )
    })

    test('ignores duplicate occurrences', () => {
      expect(parseStringToTime('2w 4d 12w 5h 15m 50s 10h 3d')).toBe(
        2 * UNIT_MULTIPLIIERS.w +
          4 * UNIT_MULTIPLIIERS.d +
          5 * UNIT_MULTIPLIIERS.h +
          15 * UNIT_MULTIPLIIERS.m +
          50 * UNIT_MULTIPLIIERS.s
      )
    })
  })
})
