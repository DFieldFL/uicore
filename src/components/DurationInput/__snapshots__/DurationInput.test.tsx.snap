// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DurationInput/> tests component tests renders with given value 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="main"
    >
      <label
        class="label"
      />
      <div
        class="bp3-input-group"
      >
        <input
          class="bp3-input"
          placeholder="Enter w/d/h/m/s/ms"
          style="padding-right: 10px;"
          type="text"
          value="1d 10h 17m 36s 789ms"
        />
      </div>
    </div>
    <div
      class="bp3-popover-wrapper helpIcon"
    >
      <span
        class="bp3-popover-target"
      >
        <span
          class="bp3-icon main"
          tabindex="0"
        >
          <svg
            height="14"
            width="14"
          />
        </span>
      </span>
    </div>
  </div>
</div>
`;

exports[`<DurationInput/> tests component tests shows help popover 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="main"
    >
      <label
        class="label"
      />
      <div
        class="bp3-input-group"
      >
        <input
          class="bp3-input"
          placeholder="Enter w/d/h/m/s/ms"
          style="padding-right: 10px;"
          type="text"
          value="1d 10h 17m 36s 789ms"
        />
      </div>
    </div>
    <div
      class="bp3-popover-wrapper helpIcon"
    >
      <span
        class="bp3-popover-target bp3-popover-open"
      >
        <span
          class="bp3-icon main"
          tabindex="0"
        >
          <svg
            height="14"
            width="14"
          />
        </span>
      </span>
      <div
        class="bp3-overlay bp3-overlay-open bp3-overlay-inline"
      >
        <div
          class="bp3-transition-container bp3-popover-appear bp3-popover-appear-active"
          style="position: absolute; top: 0px; left: 0px; transform: translate3d(NaNpx, NaNpx, 0); will-change: transform;"
        >
          <div
            class="bp3-popover"
            style="transform-origin: NaNpx bottom;"
          >
            <div
              class="bp3-popover-arrow"
            >
              <svg
                style="transform: rotate(-90deg);"
                viewBox="0 0 30 30"
              >
                <path
                  class="bp3-popover-arrow-border"
                  d="M8.11 6.302c1.015-.936 1.887-2.922 1.887-4.297v26c0-1.378-.868-3.357-1.888-4.297L.925 17.09c-1.237-1.14-1.233-3.034 0-4.17L8.11 6.302z"
                />
                <path
                  class="bp3-popover-arrow-fill"
                  d="M8.787 7.036c1.22-1.125 2.21-3.376 2.21-5.03V0v30-2.005c0-1.654-.983-3.9-2.21-5.03l-7.183-6.616c-.81-.746-.802-1.96 0-2.7l7.183-6.614z"
                />
              </svg>
            </div>
            <div
              class="bp3-popover-content"
            >
              <p
                class="font main padding padding-xlarge style style-minWidth style-minWidth-192px"
                style="min-width: 192px;"
              >
                You can use:
                <br />
                <br />
                <code>
                  <b>
                    w
                  </b>
                </code>
                  for weeks
                <br />
                <code>
                  <b>
                    d
                  </b>
                </code>
                  for days
                <br />
                <code>
                  <b>
                    h
                  </b>
                </code>
                  for hours
                <br />
                <code>
                  <b>
                    m
                  </b>
                </code>
                  for minutes
                <br />
                <code>
                  <b>
                    s
                  </b>
                </code>
                  for seconds
                <br />
                <code>
                  <b>
                    ms
                  </b>
                </code>
                 for milliseconds
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
